### DISCUSSION TITLE
Again, simple assignment does not works as expected

### DISCUSSION THREAD

**Discussion by claudiosa** [vlang/v#13900] [General]
Hi everyone,

Why is this sequence of commands is accept in the REPL of V.....

```
>>> import rand
>>> mut x := ( rand.intn(69) ) 
>>> println('x: ${x}')
x: 8
>>>  x = ( rand.intn(69) ) 
>>> println('x: ${x}')
x: 51
>>>  x = ( rand.intn(69) ) 
>>> println('x: ${x}')
x: 60
>>>  x = ( rand.intn(69) ) 
>>> println('x: ${x}')
x: 41
>>> ..................................
```
but NOT  when it is inside a code  such:

```
mut x := ( rand.intn(69) ) 
for _ in 0..4 {
  x =  ( rand.intn(69) )
  println('x: ${x}')
 }

``` 

$ v -cg run random_study.v 
/tmp/v_1317/random_study.17070852272020893212.tmp.c:13976: error: cannot convert 'struct Option_int' to 'int'
builder error: 

==================
C error. This should never happen.

This is a compiler bug, please report it using v bug file.v.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

Is it a bug? Or my mistake?

Thanks in advance

Claudio

**Comment by medvednikov** [comment]
Use ` rand.intn(1000000) or { 0 }`

**Comment by medvednikov** [comment]
You found an interesting bug in the compiler, it'll be fixed.

**Reply by claudiosa** [reply]
@medvednikov 

Yes, I think that is a bug, anyway, the notation should be improved also. However, there's the notation that was accepted (but ugly):

```
mut x := [ rand.intn(69) ? ]     // ( ) as a function
print('\n Guessing x :  ${ x[0] } '  )
x = [ rand.intn(69) ?]   // X.int()	 
print('\n Guessing x again:  ${ x[0] } '  )
	
	for _ in 0..4 {
		x  = [ rand.intn(69) ? ]    
		print('\n Guessing x againg :  ${ x[0] } ' )
		}
```
with an output:

```
$ v  run random_study.v 

 Guessing x :  44 
 Guessing x again:  61 
 Guessing x againg :  58 
 Guessing x againg :  52 
 Guessing x againg :  53 
 Guessing x againg :  50 
```
It is a workaround, but an elegant notation should be adjusted. 
Which is the reason to use it as an array to receive a random number?

Thanks once

Claudio

**Comment by trufae** [comment]
Using array syntax is not any notation. It's just how you wrote the program and its an unnecessary overcomplication

**Reply by claudiosa** [reply]
No @trufae , please look the first (thread of this topic) code compiled X interpreter and @medvednikov  's observation. There is something wrong / probablly a bug in the compiler !!!

**Comment by trufae** [comment]
@claudiosa you are writing wrong V code. you just spotted a bug in the compiler as @medvednikov said. but it's just because you are workarounding your misunderstanding on the language with invalid constructions that result on unexpected behaviour.

The correct way to handle optional returns is by using ? to escalate the error up to a handler or a panic in main, or you can just use an if-let construction or use an `or` statement to handle that case.

Here's how you should write the code. forget about parentheses or brackets:


```v
$ v
>>> import rand
>>> rand.intn(33)
error: rand.intn() returns an option, so it should have either an `or {}` block, or `?` at the end
    7 | import rand
    8 |
    9 | println(rand.intn(33))
      |              ~~~~~~~~
>>> rand.intn(33)?
2
>>>
```

The error message is clear about how to handle optional returns, it doesn't mention [] or () which are just wrong hacks because they result on nested scopes or arrays, and this is were the compiler bug lives in.

Btw, compiler bugs in V are expected because it didn't reached the 1.0 yet. So it's great time to find those bugs and report them, Thanks for spotting it, and will hopefully be fixed soon. but that's no blocker for you if you write proper V.

**Comment by claudiosa** [comment]
@trufae and @medvednikov 

My apologies for this misunderstanding. Really, a communication problem by my side.
Sorry. I will do a workaround using a call from a function with a return `(rand.intn( n ) ?)`.
Thanks.

cc
